File: sdl_bitmap_and_mouse/src/assetloader.c
/*
 * Description: functions for loading assets
 */

#include "assetloader.h"

//! Function to run all of the initial asset loading.
/*
 * @return    string    error message, if any
 */
char* LoadAssets() {

    // variable declaration
    char* err = "";

    // attempt to initialize the textures
    err = InitTextures();

    // if an error occurs, pass it back.
    if (strlen(err) > 0) {
        return err;
    }

    // everything loaded correctly
    return "";
}

//! Initializes game textures.
/*
 * @return    string    error message, if any
 */
char* InitTextures() {

    // variable declaration
    int i                    = 0;
    char img_path[256]       = {0};
    SDL_Surface* tmp_surface = NULL;

    // assemble the path string for the mouse bitmap
    sprintf(img_path, UI_GFX_LOCATION"mouse.bmp");

    // safety check, ensure the int could actually be converted
    if (strlen(img_path) < 1) {
        return "InitTextures() --> unable to assemble mouse img path";
    }

    // load the mouse graphics into memory
    tmp_surface = SDL_LoadBMP((char*) img_path);
    memset(&img_path, 0, 256*sizeof(char));
    MOUSE_GFX = SDL_CreateTextureFromSurface(RENDERER, tmp_surface);
    SDL_FreeSurface(tmp_surface);

    // ensure calloc worked
    if (!MOUSE_GFX) {
        return "InitTextures() --> unable to calloc for mouse GFX";
    }

    // attempt to calloc mem for the desert tile textures
    DESERT_TILES = (SDL_Texture**) calloc(LEVEL_TYPE_DESERT_MAX,
      sizeof(SDL_Texture*));

    // ensure calloc worked
    if (!DESERT_TILES) {
        DeleteTextures();
        return "InitTextures() --> unable to calloc for desert tiles";
    }

    // attempt to calloc mem for the grass tile textures
    GRASS_TILES = (SDL_Texture**) calloc(LEVEL_TYPE_GRASS_MAX,
      sizeof(SDL_Texture*));

    // ensure calloc worked
    if (!GRASS_TILES) {
        DeleteTextures();
        return "InitTextures() --> unable to calloc for grass tiles";
    }

    // attempt to calloc mem for the swamp tile textures
    SWAMP_TILES = (SDL_Texture**) calloc(LEVEL_TYPE_SWAMP_MAX,
      sizeof(SDL_Texture*));

    // ensure calloc worked
    if (!SWAMP_TILES) {
        DeleteTextures();
        return "InitTextures() --> unable to calloc for swamp tiles";
    }

    // Load all of the desert tiles.
    for (i = 1; i < LEVEL_TYPE_DESERT_MAX; i++) {

        // assemble the path string
        sprintf(img_path, LEVEL_TYPE_DESERT_PATH"%d.bmp", i);

        // safety check, ensure the int could actually be converted
        if (strlen(img_path) < 1) {
            DeleteTextures();
            return "InitTextures() --> unable to assemble desert img path";
        }

        // load the image into memory
        tmp_surface = SDL_LoadBMP((char*) img_path);
        memset(&img_path, 0, 256*sizeof(char));
        DESERT_TILES[i-1] = SDL_CreateTextureFromSurface(RENDERER, tmp_surface);
        SDL_FreeSurface(tmp_surface);
    }

    // Load all of the grass tiles.
    for (i = 1; i < LEVEL_TYPE_GRASS_MAX; i++) {

        // assemble the path string
        sprintf(img_path, LEVEL_TYPE_GRASS_PATH"%d.bmp", i);

        // safety check, ensure the int could actually be converted
        if (strlen(img_path) < 1) {
            DeleteTextures();
            return "InitTextures() --> unable to assemble grass img path";
        }

        // load the image into memory
        tmp_surface = SDL_LoadBMP((char*) img_path);
        memset(&img_path, 0, 256*sizeof(char));
        GRASS_TILES[i-1] = SDL_CreateTextureFromSurface(RENDERER, tmp_surface);
        SDL_FreeSurface(tmp_surface);
    }

    // Load all of the swamp tiles.
    for (i = 1; i < LEVEL_TYPE_SWAMP_MAX; i++) {

        // assemble the path string
        sprintf(img_path, LEVEL_TYPE_SWAMP_PATH"%d.bmp", i);

        // safety check, ensure the int could actually be converted
        if (strlen(img_path) < 1) {
            DeleteTextures();
            return "InitTextures() --> unable to assemble swamp img path";
        }

        // load the image into memory
        tmp_surface = SDL_LoadBMP((char*) img_path);
        memset(&img_path, 0, 256*sizeof(char));
        SWAMP_TILES[i-1] = SDL_CreateTextureFromSurface(RENDERER, tmp_surface);
        SDL_FreeSurface(tmp_surface);
    }

    return "";
}

//! Function to clear all of the loaded textures.
/*
 * @return    string    error message, if any
 *
 * TODO: consider freeing other assets here too
 */
char* DestroyAssets() {

    // variable declaration
    char* err = "";

    // attempt to clear the textures in memory
    err = DeleteTextures();

    // if an error occurs, pass it back.
    if (strlen(err) > 0) {
        return err;
    }

    // everything freed correctly
    return "";
}

//! Frees game textures from memory.
/*
 * @return    string    error message, if any
 */
char* DeleteTextures() {

    // variable declaration
    unsigned int i = 0;

    // attempt to free the mouse texture upon exit
    if (MOUSE_GFX) {
        free(MOUSE_GFX);
        MOUSE_GFX = NULL;
    }

    // If the desert textures are still loaded...
    if (DESERT_TILES) {
        for (i = 0; i < LEVEL_TYPE_DESERT_MAX; i++) {
            if (DESERT_TILES[i]) {
                free(DESERT_TILES[i]);
                DESERT_TILES[i] = NULL;
            }
        }
        free(DESERT_TILES);
        DESERT_TILES = NULL;
    }

    // If the grass textures are still loaded...
    if (GRASS_TILES) {
        for (i = 0; i < LEVEL_TYPE_GRASS_MAX; i++) {
            if (GRASS_TILES[i]) {
                free(GRASS_TILES[i]);
                GRASS_TILES[i] = NULL;
            }
        }
        free(GRASS_TILES);
        GRASS_TILES = NULL;
    }

    // If the swamp textures are still loaded...
    if (SWAMP_TILES) {
        for (i = 0; i < LEVEL_TYPE_SWAMP_MAX; i++) {
            if (SWAMP_TILES[i]) {
                free(SWAMP_TILES[i]);
                SWAMP_TILES[i] = NULL;
            }
        }
        free(SWAMP_TILES);
        SWAMP_TILES = NULL;
    }

    // all of the tile memory was freed correctly
    return "";
}

File: sdl_in_rust/src/assetloader.rs
/*
 * Description: functions for loading assets
 */

extern crate sdl2;

use std::path::PathBuf;
use sdl2::surface::*;

use std::collections::HashMap;

pub const UI_GFX_LOCATION: &str = "img/ui/";
pub const LEVEL_TYPE_DESERT_PATH: &str = "img/desert/";
pub const LEVEL_TYPE_GRASS_PATH: &str = "img/grass/";
pub const LEVEL_TYPE_SWAMP_PATH: &str = "img/swamp/";

pub const LEVEL_TYPE_DESERT_MAX: i32 = 6;
pub const LEVEL_TYPE_GRASS_MAX: i32 = 14;
pub const LEVEL_TYPE_SWAMP_MAX: i32 = 13;

//
// Assets
//
pub struct Assets<'a> {
    tcs: HashMap<&'static str, sdl2::render::TextureCreator<sdl2::video::WindowContext>>,
    textures: HashMap<&'static str, sdl2::render::Texture<'a>>,
}

impl<'a> Assets<'a> {
    pub fn new() -> Assets<'a> {
        Assets {tcs: HashMap::new(), textures: HashMap::new()}
    }

    pub fn init_textures(
        &'a mut self,
        canvas: &sdl2::render::Canvas<sdl2::video::Window>
        ) -> Result<(), String> {

        self.tcs.insert("mouse_gfx_tc", canvas.texture_creator());

        let path_to_mouse_gfx: PathBuf = [UI_GFX_LOCATION, "mouse.bmp"].iter().collect();

        match path_to_mouse_gfx.canonicalize() {
            Err(e) => return Err(e.to_string()),
            Ok(path) => {

                match Surface::load_bmp(path) {
                    Err(e) => return Err(e.to_string()),
                    Ok(mouse_surface) => {

                         match self.tcs["mouse_gfx_tc"].create_texture_from_surface(mouse_surface) {
                             Err(e) => return Err(e.to_string()),
                             Ok(mouse_gfx) => {
                                 self.textures.insert("mouse_gfx", mouse_gfx);
                             },
                         };
                    },
                };
            },
        };

        // Load all of the desert tiles.
        for i in 1..LEVEL_TYPE_DESERT_MAX {

            let path: PathBuf = [LEVEL_TYPE_DESERT_PATH, &format!("{}",i), ".bmp"].iter().collect();

            let path_canon = match path.canonicalize() {
                Err(e) => return Err(e.to_string()),
                Ok(p) => p,
            };

            //tmp_surface = SDL_LoadBMP((char*) img_path);
            //if (!tmp_surface) {
            //    return "InitTextures() --> unable to open file in: "LEVEL_TYPE_DESERT_PATH;
            //}

            //// load the image into memory
            //memset(&img_path, 0, 256*sizeof(char));
            //DESERT_TILES[i-1] = SDL_CreateTextureFromSurface(RENDERER, tmp_surface);

            //SDL_FreeSurface(tmp_surface);
            //tmp_surface = NULL;
        }

        // Load all of the grass tiles.
        for i in 1..LEVEL_TYPE_GRASS_MAX {

            let path: PathBuf = [LEVEL_TYPE_GRASS_PATH, &format!("{}",i), ".bmp"].iter().collect();

            let path_canon = match path.canonicalize() {
                Err(e) => return Err(e.to_string()),
                Ok(p) => p,
            };

            //tmp_surface = SDL_LoadBMP((char*) img_path);
            //if (!tmp_surface) {
            //    return "InitTextures() --> unable to open file in: "LEVEL_TYPE_GRASS_PATH;
            //}

            //// load the image into memory
            //memset(&img_path, 0, 256*sizeof(char));
            //GRASS_TILES[i-1] = SDL_CreateTextureFromSurface(RENDERER, tmp_surface);

            //SDL_FreeSurface(tmp_surface);
            //tmp_surface = NULL;
        }

        // Load all of the swamp tiles.
        for i in 1..LEVEL_TYPE_SWAMP_MAX {

            let path: PathBuf = [LEVEL_TYPE_SWAMP_PATH, &format!("{}",i), ".bmp"].iter().collect();

            let path_canon = match path.canonicalize() {
                Err(e) => return Err(e.to_string()),
                Ok(p) => p,
            };

            //tmp_surface = SDL_LoadBMP((char*) img_path);
            //if (!tmp_surface) {
            //    return "InitTextures() --> unable to open file in: "LEVEL_TYPE_SWAMP_PATH;
            //}

            //// load the image into memory
            //memset(&img_path, 0, 256*sizeof(char));
            //SWAMP_TILES[i-1] = SDL_CreateTextureFromSurface(RENDERER, tmp_surface);

            //SDL_FreeSurface(tmp_surface);
            //tmp_surface = NULL;
        }

        Ok(())
    }
}
